pkg/networkservice/common/clienturl/client.go

```diff
@@ -55,36 +56,58 @@ func NewClient(ctx context.Context, clientFactory func(ctx context.Context, cc g
 }
 
 func (u *clientURLClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
-	if err := u.init(); err != nil {
+	if err := u.init(ctx); err != nil {
 		return nil, err
 	}
 	return u.client.Request(ctx, request)
 }
 
 func (u *clientURLClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
-	if err := u.init(); err != nil {
+	if err := u.init(ctx); err != nil {
 		return nil, err
 	}
 	return u.client.Close(ctx, conn)
 }
 
-func (u *clientURLClient) init() error {
+func (u *clientURLClient) init(requestCtx context.Context) error {
 	u.initOnce.Do(func() {
 		clientURL := clienturlctx.ClientURL(u.ctx)
 		if clientURL == nil {
 			u.dialErr = errors.New("cannot dial nil clienturl.ClientURL(ctx)")
 			return
 		}
+
+		ctx, cancel := context.WithCancel(u.ctx)
+		errCh := make(chan error, 1)
 		var cc *grpc.ClientConn
-		cc, u.dialErr = grpc.DialContext(u.ctx, grpcutils.URLToTarget(clientURL), u.dialOptions...)
-		if u.dialErr != nil {
+		go func() {
+			var dialErr error
+			cc, dialErr = grpc.DialContext(ctx, grpcutils.URLToTarget(clientURL), u.dialOptions...)
+			if dialErr != nil {
+				errCh <- dialErr
+				return
+			}
+			errCh <- nil
+		}()
```

Comment:
Why do we need a goroutine here?

Reply:
To exit on timeout. We cannot use context with deadline here

Reply:
Fixed